From ae0b700a856c0ae460d271bb50dccfaae84dbcab Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Sat, 12 Sep 2020 00:09:07 +0200
Subject: [PATCH] hashmap: make sure to initialize shared hash key atomically
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1913423
Bug: https://github.com/systemd/systemd/issues/17007
Origin: upstream, https://github.com/systemd/systemd/commit/ae0b700a856c0ae460d271bb50dccfaae84dbcab

if we allocate a bunch of hash tables all at the same time, with none
earlier than the other, there's a good chance we'll initialize the
shared hash key multiple times, so that some threads will see a
different shared hash key than others.

Let's fix that, and make sure really everyone sees the same hash key.

Fixes: #17007
---
 src/basic/hashmap.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

--- a/src/basic/hashmap.c
+++ b/src/basic/hashmap.c
@@ -20,6 +20,7 @@
 ***/
 
 #include <errno.h>
+#include <pthread.h>
 #include <stdint.h>
 #include <stdlib.h>
 #include <string.h>
@@ -38,7 +39,6 @@
 #include "util.h"
 
 #if ENABLE_DEBUG_HASHMAP
-#include <pthread.h>
 #include "list.h"
 #endif
 
@@ -213,7 +213,6 @@ assert_cc(DIRECT_BUCKETS(struct set_entr
  * a handful of directly stored entries in a hashmap. When a hashmap
  * outgrows direct storage, it gets its own key for indirect storage. */
 static uint8_t shared_hash_key[HASH_KEY_SIZE];
-static bool shared_hash_key_initialized;
 
 /* Fields that all hashmap/set types must have */
 struct HashmapBase {
@@ -747,6 +746,10 @@ static void reset_direct_storage(Hashmap
         memset(p, DIB_RAW_INIT, sizeof(dib_raw_t) * hi->n_direct_buckets);
 }
 
+static void shared_hash_key_initialize(void) {
+        random_bytes(shared_hash_key, sizeof(shared_hash_key));
+}
+
 static struct HashmapBase *hashmap_base_new(const struct hash_ops *hash_ops, enum HashmapType type HASHMAP_DEBUG_PARAMS) {
         HashmapBase *h;
         const struct hashmap_type_info *hi = &hashmap_type_info[type];
@@ -770,10 +773,8 @@ static struct HashmapBase *hashmap_base_
 
         reset_direct_storage(h);
 
-        if (!shared_hash_key_initialized) {
-                random_bytes(shared_hash_key, sizeof(shared_hash_key));
-                shared_hash_key_initialized= true;
-        }
+        static pthread_once_t once = PTHREAD_ONCE_INIT;
+        assert_se(pthread_once(&once, shared_hash_key_initialize) == 0);
 
 #if ENABLE_DEBUG_HASHMAP
         h->debug.func = func;
--- a/meson.build
+++ b/meson.build
@@ -2300,6 +2300,7 @@ if conf.get('ENABLE_HWDB') == 1
                          'src/libsystemd/sd-hwdb/hwdb-internal.h',
                          include_directories : includes,
                          link_with : [libudev_static],
+                         dependencies : [threads],
                          install_rpath : udev_rpath,
                          install : true,
                          install_dir : rootbindir)
